### ISSUE TITLE
Allow invoking maven commands from command palette.

### ISSUE THREAD

**Issue by rbingham** [Microsoft/vscode-maven#31]
Currently, to run maven commands, you have to use the maven tree. It would be great to be able to run these commands from the command palette.

**Comment by Eskibear** [comment]
@rbingham We once supported that, and disabled it later.
Suppose you'd like to trigger `mvn install` for one of the projects/modules, you first trigger some "Maven: Install" command in command palette, then you have to specify for which project/module you'd like to run the command.

I myself also prefer using keyboard rather than moving and clicking the mouse. But we have to specify the target project if there are more than one. I used to provide a QuickPick list of all projects for users to select, but it makes things a little bit complicated. What do you think?

**Comment by rbingham** [comment]
I can understand that it can be a bit complicated, but you could start with having them pick the project that they want to run the command, then pick which command they want (this should include any custom commands that they have for the specific project). You could also cache the most used commands/projects, and have those appear at the top of the list so it is a little easier to navigate.

So, the options would be:
open command palette -> run maven -> select project -> select command

That may be too many steps, but I think that it is better than having to use the mouse (I like to keep the file explorer closed to save on code real estate, so for me there are more steps to run maven than the above)

**Comment by Eskibear** [comment]
Yes, this makes sense. I used to cache the list of maven projects/modules. At that time, I registered "command" for each maven command, and the workflow looks like:
```
command palette -> select a maven command (from a long list of commands) -> select project
```

> open command palette -> run maven -> select project -> select command

I think the workflow you proposed is ok. The team will have a discussion on this feature.

**Comment by rbingham** [comment]
Another thing that would be useful would be to be able to assign a keyboard shortcut to a project/command combo for the most common run commands.

**Comment by Eskibear** [comment]
That's why we provide `custom goals` command. And yes that might be useful for some users, but we need to avoid over design. The following workflow also looks good to me:

> command palette -> run maven -> project -> custom goals -> the first one is the latest used.

Alright when I'm typing here, it does feel a little bit complicated. How about this:

> command palette -> maven history -> press <enter> (most recently used command on the top)

@rbingham What do you think?

**Comment by Eskibear** [comment]
> command palette -> maven history -> press enter (most recently used command on the top)

I prefer this one. Example: 
![image](https://user-images.githubusercontent.com/2351748/38462268-bb7d8984-3b16-11e8-9b88-8c5724a06c7f.png)

Each item:
```
[recently used command] [project name]
[pom file absolute path]
```

But it's overlap with the command `custom goals ...` in part of the functionality. 
Previously we right click -> `custom goals...`, then there is a list of historical commands, and 2 options `New` and `Edit`. If there is a new command `maven.history`, I think this part should be removed. The new  workflow of `custom goals ...` should be simply as following:

> Right click  ->  custom goals ... -> input > enter

**Comment by Eskibear** [comment]
The feature would be separated to several tasks:

1. Simplify workflow of `maven.goal.custom`, allowing to input new custom goals only.
1. Keep a record of LRU goals, with information of goals, project name, pom file path.
1. Add a new command `maven.goal.history` which can be triggered in command palette.
    * If triggered in command palette, it list all LRU commands.
    * If triggered by right-click of a project, it filters LRU commands only for the project.

It would be a breaking change, and we should also help users not to lose their historical custom goals. 🤔

**Comment by Eskibear** [comment]
shipped in v0.8.0

### supported workflow:
Keyboard invocation:
> command palette -> maven execute -> select project -> select command

Run historical commands:
> command palette -> maven history -> press enter (most recently used command on the top)
